frontend/src/metabase/query_builder/actions/core/updateQuestion.ts

```diff
@@ -129,17 +131,19 @@ export const updateQuestion = (
 
       // When the dataset query changes, we should change the question type,
       // to start building a new ad-hoc question based on a dataset
-      if (newQuestion.isDataset()) {
+      if (newQuestion.type() === "model" || newQuestion.type() === "metric") {
         newQuestion = newQuestion.setType("question");
         dispatch(onCloseQuestionInfo());
       }
     }
 
-    // This scenario happens because the DatasetQueryEditor converts the dataset/model question into a normal question
+    // This scenario happens because the DatasetQueryEditor converts the model/metric question into a normal question
     // so that its query is shown properly in the notebook editor. Various child components of the notebook editor have access to
     // this `updateQuestion` action, so they end up triggering the action with the altered question.
     if (queryBuilderMode === "dataset" && !newQuestion.isDataset()) {
-      newQuestion = newQuestion.setType("model");
+      const routing = getRouting(getState());
+      const type = getCardTypeFromLocation(routing.locationBeforeTransitions);
+      newQuestion = newQuestion.setType(type);
```

Comment:
This is an extension to the existing "hack".
Now `queryBuilderMode === "dataset"` is shared by both models and metrics.
URL is the only place where I could find information on what type of entity are we dealing with.

Alternative solutions I considered:
- remove the original hack - [this is the best option, but also the most difficult](https://metaboat.slack.com/archives/C505ZNNH4/p1707400291390129)
- add one more piece of information to `state.qb.uiControls` indicating what kind of entity did we start with - simple, but might have edge cases
- introduce a new `queryBuilderMode` - would require a lot of changes

Reply:
I managed to get rid of this: c1d3db383613b4bdebfdbfcfb1620b144d3e1503